table_cleanup.go

```diff
@@ -0,0 +1,97 @@
+package kbucket
+
+import (
+	"context"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+func (rt *RoutingTable) cleanup() {
+	validatePeerF := func(p peer.ID) bool {
+		queryCtx, cancel := context.WithTimeout(rt.ctx, rt.peerValidationTimeout)
+		defer cancel()
+		return rt.peerValidationFnc(queryCtx, p)
+	}
+
+	cleanupTickr := time.NewTicker(rt.tableCleanupInterval)
+	defer cleanupTickr.Stop()
+	for {
+		select {
+		case <-cleanupTickr.C:
+			ps := rt.peersToValidate()
+			for _, pinfo := range ps {
+				// TODO This is racy
+				// A peer could disconnect immediately after we validate it & would thus be in missing state again
+				// which means we would wrongly mark it as active here. The longer term solution is to
+				// handle all peer related events in a single event loop in the RT or more fingrained locking at bucket/peer level.
+				// See https://github.com/libp2p/go-libp2p-kbucket/issues/60
+				if validatePeerF(pinfo.Id) {
+					rt.tabLock.Lock()
+					// add it back/mark it as active ONLY if it is still in the RT
+					// to avoid adding it back if it's been marked as dead
+					i := rt.bucketIdForPeer(pinfo.Id)
+					if rt.buckets[i].getPeer(pinfo.Id) != nil {
+						log.Infof("successfully validated missing peer=%s, marking it as active", pinfo.Id)
+						rt.addPeer(pinfo.Id)
+					}
+					rt.tabLock.Unlock()
+					continue
+				}
+
+				// peer does not seem to be alive, let's try to replace it
+				// evict missing peer & request replacement ONLY if it's NOT marked as active to avoid removing a peer that connected after
+				// the failed validation
+				rt.tabLock.Lock()
+				i := rt.bucketIdForPeer(pinfo.Id)
+				p := rt.buckets[i].getPeer(pinfo.Id)
+				if p != nil && p.State != PeerStateActive {
+					log.Infof("failed to validate missing peer=%s, evicting it from the RT & requesting a replace", pinfo.Id)
+					rt.removePeer(pinfo.Id)
+				}
+				rt.tabLock.Unlock()
+			}
+		case <-rt.ctx.Done():
+			return
+		}
+	}
+}
+
+// replaces a peer using a valid peer from the replacement cache
+func (rt *RoutingTable) startPeerReplacement() {
```

Comment:
Do we need another goroutine or can we just have a single "run" goroutine?

Reply:
Replacement is also requested by the DHT in addition to the cleanup routine. What do you have in mind when you say have a single "run" go-routine ?

Reply:
I mean merge the cleanup and replacement goroutines to reduce the number fo components updating the routing table. But it's not really an important change.